### PR TITLE
[Review] Hash Join Optimization: Hash entire row instead of just the first column

### PR DESCRIPTION
**Pull request by jrhemstad**
- Improves the existing hash join implementation by hashing the entire row of the table instead of just the first column
-  Modifies the `concurrent_unordered_multimap::insert` and `concurrent_unordered_multimap::find` functions to optionally accept a pre-computed hash value for the key being inserted/found. This pre-computed hash value will be used to determine the location of the key in the table.
- Updated the `concurrent_unordered_multimap::insert` function to detect when the map is full and break out of execution to avoid an infinite loop.
-  The `build_hash_table` kernel was updated to compute the hash value of each row in the build table and insert the hash value as the key in the hash table.
-  The `compute_join_output_size` and `probe_hash_table` kernels were updated so that when probing the hash table, it first checks that the hash value of the probe row equals the hash value stored as the key in the hash table before doing the full row equality comparison.
-  An unrelated change was removing the host-side tests of the `concurrent_unordered_map` and `concurrent_unordered_multimap` member functions to remove warnings about using a host function inside of a device function. These classes are moving towards being device-only classes.
- Cleans up obsolete code related to the deprecated `gdf_join_result_type`
- Adds new unit tests to test very large inputs 
- Updates `size_type` used inside of the hash join implementation from `int` to `int64_t` to avoid integer overflow issues with large inputs
- Added an error code to the build_hash_table kernel. Added checking if the insert() fails and setting the GDF_HASH_TABLE_INSERT_FAILURE error code accordingly.

### CODE REVIEW

**Comment by scopatz** [comment]
This has some conflicts now

File: src/tests/groupby/groupby-kernels-test.cu [hunk 1]
```diff
@@ -383,50 +383,7 @@ typedef ::testing::Types<
 
   TYPED_TEST_CASE(GroupByTest, Implementations);
 
-TYPED_TEST(GroupByTest, DISABLED_AggregationTestHost)
```
Comment (scopatz) [inline src/tests/groupby/groupby-kernels-test.cu]: Hmm I don't think it OK to remove these tests. Can you please add them back and get them working? Thanks!

File: src/tests/join/join-tests.cu [hunk 2]
```diff
@@ -560,14 +558,15 @@ typedef ::testing::Types<
 
 TYPED_TEST_CASE(JoinTest, Implementations);
 
-TYPED_TEST(JoinTest, ExampleTest)
+TYPED_TEST(JoinTest, DISABLED_DebugTest)
```
Comment (scopatz) [inline src/tests/join/join-tests.cu]: Why is this test being disabled?

File: src/tests/groupby/groupby-kernels-test.cu (hunk 1 above)
Comment (jrhemstad) [reply src/tests/groupby/groupby-kernels-test.cu]: These tests test the host-side implementation of the `concurrent_unordered_multimap` and `concurrent_unordered_map` classes insert functions.  

I'm in the process of updating these classes to no longer have host-side implementations of these functions as they require the class to use managed memory, which complicates the custom memory manager that is being developed. 

Furthermore, there is no good reason to have host-side implementations of these functions and maintaining them just adds unnecessary work and code bloat.

Comment (scopatz) [reply src/tests/groupby/groupby-kernels-test.cu]: That makes sense, but could you add device side tests to replace them?

File: src/tests/join/join-tests.cu (hunk 2 above)
Comment (jrhemstad) [reply src/tests/join/join-tests.cu]: This test is disabled because it's a test I use for debug purposes. It has small input sizes and has extensive debug printing enabled. Having it enabled by default would add excessive clutter to the unit test output. I'll add a comment to clarify this.

Comment (scopatz) [reply src/tests/join/join-tests.cu]: OK! Thanks

File: src/tests/groupby/groupby-kernels-test.cu (hunk 1 above)
Comment (jrhemstad) [reply src/tests/groupby/groupby-kernels-test.cu]: See `/src/tests/hash_map/map-test.cu`.

However, this brings up a good question related to unit testing. For example, should unit tests just test the top-level libgdf APIs in `functions.h`? Or should there also be unit testing for the classes/functions/kernels that are used to implement the top level API? 

In this case, we're talking about tests for the lower-level implementation classes, which are turning out to be a lot of work to maintain as implementations are updated.

I'll pose the question in Slack.

Comment (scopatz) [reply src/tests/groupby/groupby-kernels-test.cu]: Thanks! Yeah it would be great to get some clarity on what precisely needs to be tested.

Comment (felipeblazing) [reply src/tests/groupby/groupby-kernels-test.cu]: We are trying, with lukewarm results, to create unit tests for most functions we make. So not just the top level apis but the lower level functions that get leveraged as well. If we only test top level apis those are not unit tests. Those are integration tests. And those are great but unit tests really should test only what happens in the function and mock up whatever happens outside of it.

File: src/hashmap/concurrent_unordered_multimap.cuh [hunk 3]
```diff
@@ -412,40 +456,72 @@ public:
...
+            hash_tbl_idx = (hash_tbl_idx+1)%hashtbl_size;
+
+            attempt_counter++;
+            if( attempt_counter > hashtbl_size)
             {
-                if ( m_equal( unused_key, tmp_it->first ) ) {
-                    hashtbl_values[hash_tbl_idx] = thrust::make_pair( x.first, x.second );
-                    it = tmp_it;
-                }
+              printf("Attempted to insert to multimap but the map is full!\n");
```
[the remark this answers is not in the document]
Comment (jrhemstad) [reply src/hashmap/concurrent_unordered_multimap.cuh]: Currently nothing will happen other than printing the error and exiting the loop, which is better than what would currently happen---infinite loop. 

Technically the hashmap should never be full by construction, but there was a bug where the computation to determine the size of the hash table was overflowing an `int` and therefore was ending up smaller than it was supposed to be. So this is an attempt to at least do something other than cause an infinite loop. We can't really check for `end()` and return an error, because `insert` is called in a device kernel, and we can't return an error code from a kernel.

File: src/join/hash/join_compute_api.h [hunk 4]
```diff
@@ -30,204 +31,299 @@
...
-                                     probe_column, 
-                                     probe_table.get_column_length(), 
-                                     output_l_ptr,
-                                     output_r_ptr,
-                                     d_global_write_index, 
-                                     h_join_output_size,
-                                     flip_results);
-
-  CUDA_RT_CALL(cudaDeviceSynchronize());
+  CUDA_TRY( cudaMalloc(&d_global_write_index, sizeof(size_type)) );
```
[the remark this answers is not in the document]
Comment (jrhemstad) [reply src/join/hash/join_compute_api.h]: @harrism has already gone through most of libgdf and replaced all the cudaMallocs with wrappers. His PR should be coming through soon.

File: src/hashmap/concurrent_unordered_multimap.cuh (hunk 3 above)
[the remark this answers is not in the document]
Comment (nsakharnykh) [reply src/hashmap/concurrent_unordered_multimap.cuh]: I suggest to have a threshold for the max number of hash checks, and if we exceed that, we store the row index in some preallocated buffer in system memory. Then outside the kernel we print the error, indicating the row number, which could be later extended to passing it properly all the way up to PyGDF. I just don’t like doing a printf from within a kernel

[the remark this answers is not in the document]
Comment (jrhemstad) [reply src/hashmap/concurrent_unordered_multimap.cuh]: How about I pass a `gdf_error *` into the kernel, and if a thread detects that the hashmap is full, then it can set it to `GDF_HASH_TABLE_FULL`. We can then check the status of the `gdf_error` after the build kernel.

File: src/join/hash/join_compute_api.h [hunk 5]
```diff
@@ -30,204 +31,305 @@
...
-  
-  // If the output size is zero, return immediately
-  if(0 == h_join_output_size){
+  size_type estimated_join_output_size{0};
+  error = estimate_join_output_size<join_type, multimap_type>(build_table, probe_table, *hash_table, &estimated_join_output_size);
+
+  if(GDF_SUCCESS != error)
+    return error;
+
+  // If the estimated output size is zero, return immediately
```
[the remark this answers is not in the document]
Comment (nsakharnykh) [reply src/join/hash/join_compute_api.h]: Probably an easy fix would be to set the estimated output size to at least 1

File: src/tests/join/join-tests.cu [hunk 6]
```diff
@@ -644,3 +613,50 @@ TYPED_TEST(JoinTest, RightColumnsBigger)
     EXPECT_EQ(reference_result[i], gdf_result[i]);
   }
 }
+
+// TODO Add test for inputs with zero rows
```
Comment (nsakharnykh) [inline src/tests/join/join-tests.cu]: Should add tests to check the zero matching rate (either no matches, or an empty dataset?)

File: src/join/hash/join_compute_api.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (jrhemstad) [reply src/join/hash/join_compute_api.h]: `error` just happens to be the `gdf_error` variable that I use to store the return type. `error` is initialized with `GDF_SUCCESS`, so if no other errors occur, then returning it here will return `GDF_SUCCESS`.

File: src/tests/join/join-tests.cu (hunk 6 above)
Comment (jrhemstad) [reply src/tests/join/join-tests.cu]: Done.
